src/relax/op/memory/view.cc

```diff
@@ -334,13 +334,12 @@ Expr LegalizeView(const BlockBuilder& bb, const Call& call) {
     relative_byte_offset = relax::PrimValue::Int64(0);
   }
 
-  StructInfoDeriveFunc infer_sinfo_env_func;
-  infer_sinfo_env_func = EnvFunc::Get("tvm.relax.struct_info.infer_view_sinfo");
-  auto runtime_view_sinfo = FuncStructInfo::OpaqueFunc(infer_sinfo_env_func, true);
-
-  ExternFunc runtime_view_func("runtime.TVMArrayCreateView", runtime_view_sinfo);
+  if (shape.same_as(call->args[1]) && dtype.same_as(call->args[2]) &&
+      relative_byte_offset.same_as(call->args[3])) {
+    return call;
+  }
 
-  return Call(runtime_view_func, {data, shape, dtype, relative_byte_offset});
+  return Call(call->op, {data, shape, dtype, relative_byte_offset});
```

Comment:
This only does the inference of `void` type args and leave the lowering to the later pass.

Reply:
I don't think we want to do the inference of the shape/dtype prior to lowering, because it could result in unexpected StructInfo inference later on.

Suppose we have `R.memory.view(arg, shape=[16])`.  This returns a view into the first 16 elements of `arg`, without changing the dtype.  If an `IRModule` pass updates the datatype of `arg`, then that new datatype should also propagate to the view.  However, legalizing it to `R.memory.view(arg, shape=[16], dtype="float16")` would return a view into `arg`, interpreting the first 32 bytes as if they were `"float16"`.  Now, if an `IRModule` pass updates the datatype of `arg`, the view would still be `"float16"`.  To avoid this issue, the unknown arguments shouldn't be filled in until the lowering is about to occur.

What if we were to remove `.set_attr<FLegalize>` altogether, and only have `.set_attr<FLowerBuiltin>`?  That way, we preserve the `R.memory.view` as-is until it is ready to be lowered.  The `LegalizeOps` pass would then be a no-op for `R.memory.view`, and only the `LowerRuntimeBuiltin` pass would change it at all.